contracts/FeralfileEnglishAuction.sol

```diff
@@ -0,0 +1,334 @@
+        uint256 amount_,
+        uint256 timestamp_
+    ) internal pure {
+        // make sure the auction exist
+        require(auction_.id > 0, "FeralfileEnglishAuction: auction not found");
+        // check if auction is on going
+        require(
+            auction_.startAt <= timestamp_ && timestamp_ <= auction_.endAt,
+            "FeralfileEnglishAuction: auction not started or ended"
+        );
+        // check if auction is not settled
+        require(
+            !auction_.isSettled,
+            "FeralfileEnglishAuction: auction already settled"
+        );
+        // check if current winner not equal to current bidder (optional)
+        require(
+            highestBid_.bidder != bidder_,
+            "FeralfileEnglishAuction: bidder is the current winner"
+        );
+
+        if (highestBid_.amount > 0) {
+            // check bidding amount follow the minimum increment
+            uint256 minIncreaseAmount_ = (highestBid_.amount *
+                auction_.minIncreaseFactor) / 100;
+            if (minIncreaseAmount_ < auction_.minIncreaseAmount) {
+                minIncreaseAmount_ = auction_.minIncreaseAmount;
+            }
+            require(
+                amount_ >= highestBid_.amount + minIncreaseAmount_,
+                "FeralfileEnglishAuction: bid amount should follow the minimum increment"
+            );
+        } else {
+            // check bidding amount follow the minimum price
+            require(
+                amount_ >= auction_.minPrice,
+                "FeralfileEnglishAuction: bid amount should be greater than minimum price"
+            );
+        }
+    }
+
+    function placeBid(uint256 auctionID_) external payable {
+        // get the auction object
+        Auction memory auction_ = auctions[auctionID_];
+        // get the last auction bid object
+        Bid memory highestBid_ = highestBids[auctionID_];
+        // verify the bid is valid
+        _validateBidRequest(
+            auction_,
+            highestBid_,
+            msg.sender,
+            msg.value,
+            block.timestamp
+        );
+        // transfer last winning bid amount to last bidder if fromFeralFile is false
+        if (!highestBid_.fromFeralFile) {
+            payable(highestBid_.bidder).transfer(highestBid_.amount);
+        }
+        // replace the new bid to bid map
+        highestBids[auctionID_] = Bid({
+            bidder: msg.sender,
+            amount: msg.value,
+            fromFeralFile: false
+        });
+        // if the gap to end time is lower than threshold, update the auction end time to now() + extend duration
+        if (auction_.endAt - block.timestamp <= auction_.extendThreshold) {
+            auctions[auctionID_].endAt =
+                block.timestamp +
+                auction_.extendDuration;
+        }
+        // emit new bid event
+        emit NewBid(auctionID_, msg.sender, msg.value, false);
+    }
+
+    function placeSignedBid(
```

Comment:
I think we could create a private function for `placeBid()` with flag `isFromFeralFile` additionally so that we could reuse most of code between `placeBid()` and `placeSignedBid`. The difference is only validate the signature.